# Lab 2
With Adam Rustad, Ethan Bang, Cassidy Grover, Joshua Yamada

In [39]:
import itertools
import pandas as pd
import numpy as np

# List out all possible rolls, in this case for 3d6
rolls = list(itertools.product(range(1, 7), repeat=3))

# Find the sum of all possible rolls
sums = np.array([sum(roll) for roll in rolls])

# Find the proportion of those sums that are greater than or equal to the target, in this case 14
np.mean(sums >= 14)

0.16203703703703703

### Scenario a

In [40]:
def scenario_a(m, n, x, debug = False):
  """Returns the probability that the sum of m n-sided dice is greater than or
  equal to x"""

  #m is amount of dice (max 4)
  if m > 4:
    return RuntimeError("Too many dice")

  #N is the number sides max 20
  if n > 20:
    return RuntimeError("Too many sides")

  #Probability (n + 1 to include that number)
  rolls = list(itertools.product(range(1, n+1), repeat=m))

  # Find the sum of all possible rolls
  sums = np.array([sum(roll) for roll in rolls])

  if debug:
    print(f"The Sums: {sums}")


  # Find the proportion of those sums that are greater than or equal to the target
  return np.mean(sums >= x) #True and false array


In [41]:
#Test function
scenario_a(3, 6, 14, True)

The Sums: [ 3  4  5  6  7  8  4  5  6  7  8  9  5  6  7  8  9 10  6  7  8  9 10 11
  7  8  9 10 11 12  8  9 10 11 12 13  4  5  6  7  8  9  5  6  7  8  9 10
  6  7  8  9 10 11  7  8  9 10 11 12  8  9 10 11 12 13  9 10 11 12 13 14
  5  6  7  8  9 10  6  7  8  9 10 11  7  8  9 10 11 12  8  9 10 11 12 13
  9 10 11 12 13 14 10 11 12 13 14 15  6  7  8  9 10 11  7  8  9 10 11 12
  8  9 10 11 12 13  9 10 11 12 13 14 10 11 12 13 14 15 11 12 13 14 15 16
  7  8  9 10 11 12  8  9 10 11 12 13  9 10 11 12 13 14 10 11 12 13 14 15
 11 12 13 14 15 16 12 13 14 15 16 17  8  9 10 11 12 13  9 10 11 12 13 14
 10 11 12 13 14 15 11 12 13 14 15 16 12 13 14 15 16 17 13 14 15 16 17 18]


0.16203703703703703

### Scenario B

In [42]:
def scenario_b(m, n, x, debug = False):
  """Determining the probability of rolling m dice, each with n sides, and having at least one die
  show a value greater than or equal to x."""


  #m is amount of dice (max 4)
  if m > 4:
    return RuntimeError("Too many dice")

  #N is the number sides max 20
  if n > 20:
    return RuntimeError("Too many sides")

  #Probability (n + 1 to include that number)
  rolls = list(itertools.product(range(1, n+1), repeat=m))

  if debug:
    print(f"The rolls: {rolls}")

  # Find the sum of all possible rolls
  roll_seperated = np.array([roll for roll in rolls])

  if debug:
    print(f"The roll_seperated: {roll_seperated}")

  # Find the proportion of those sums that are greater than or equal to the target
  num_success = 0
  for roll in roll_seperated:
    if (roll >= x).any():
      num_success += 1

  return num_success/len(roll_seperated)


In [43]:
(np.array([1,3,1]) > 4).any()

False

In [44]:
# Testing if there would be an instance of rolling a 7 or greater with 3x 6-side die; it should result as impossible (0.0)
scenario_b(3, 6, 7, False)

0.0

In [45]:
# Testing the probability of an instance of rolling a 6 or greater with 3x 6-side die
scenario_b(3, 6, 6, False)

0.4212962962962963

In [47]:
# Testing the probability of an instance of rolling a 0 or greater with 3x 6-side die; it should result in 1.0 (you can't roll lower than 1 on any roll)
scenario_b(3, 6, 0, False)

1.0

In [49]:
# Option 1
# Evaluating the probability of rolling 1d20 and get a result of 17 or more
scenario_a(1, 20, 17)

0.2

In [52]:
# Option 2
# evaluating the probability of rolling 3d4 and get a result of 10 or more
scenario_a(3, 4, 10)

0.15625

In [54]:
# Option 3
# evaluating the probability of rolling 4d6 and have at least one dice show a 6
scenario_b(4, 6, 6)

0.5177469135802469

The first two options utilize the same scenario as Scenario A where we are calculating the probabilities of the sum of our rolls. Option 3 utilizes the same scenario as Scenario B where we calculate the probability of rolling at least one instance where the number sought after or greater is rolled. From the given options, the probabilities of option 1, 2, and 3 are respectively about 0.200, 0.156, 0.518. Looking at these probabilities, Option 3 clearly has the highest chance of occuring because there only has to be one occurrance of a 6 and in all combinations of rolling 4d6 a 6 shows up more than half the time. Option 1 is similar to how Option 3 functions, but since only one dice is rolled the probability is equal to the probability of rolling that number plus the probability of rolling each number greater than it. The least likely to happen is Option 2 because the maximum sum that one could roll with 3d4 is 12 meaning that as the required sum nears 12 the probability of rolling 10 or greater decreases because there are less combinations that satisfy the requirement.